File: libs/services/src/os/StatsLogEventWrapper.cpp
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <android/os/StatsLogEventWrapper.h>

#include <binder/Parcel.h>
#include <binder/Parcelable.h>
#include <binder/Status.h>
#include <utils/RefBase.h>
#include <vector>

using android::Parcel;
using android::Parcelable;
using android::status_t;
using std::vector;

namespace android {
namespace os {

StatsLogEventWrapper::StatsLogEventWrapper(){};

status_t StatsLogEventWrapper::writeToParcel(Parcel* out) const {
  // Implement me if desired. We don't currently use this.
  ALOGE(
      "Cannot do c++ StatsLogEventWrapper.writeToParcel(); it is not "
      "implemented.");
  (void)out;  // To prevent compile error of unused parameter 'in'
  return UNKNOWN_ERROR;
};

status_t StatsLogEventWrapper::readFromParcel(const Parcel* in) {
  status_t res = OK;
  if (in == NULL) {
    ALOGE("statsd received parcel argument was NULL.");
    return BAD_VALUE;
  }
  if ((res = in->readInt32(&mTagId)) != OK) {
    ALOGE("statsd could not read tagId from parcel");
    return res;
  }
  if ((res = in->readInt64(&mElapsedRealTimeNs)) != OK) {
    ALOGE("statsd could not read elapsed real time from parcel");
    return res;
  }
  if ((res = in->readInt64(&mWallClockTimeNs)) != OK) {
    ALOGE("statsd could not read wall clock time from parcel");
    return res;
  }
  int numWorkChain = 0;
  if ((res = in->readInt32(&numWorkChain)) != OK) {
    ALOGE("statsd could not read number of work chains from parcel");
    return res;
  }
  if (numWorkChain > 0) {
    for (int i = 0; i < numWorkChain; i++) {
      int numNodes = 0;
      if ((res = in->readInt32(&numNodes)) != OK) {
        ALOGE(
            "statsd could not read number of nodes in work chain from parcel");
        return res;
      }
      if (numNodes == 0) {
        ALOGE("empty work chain");
        return BAD_VALUE;
      }
      WorkChain wc;
      for (int j = 0; j < numNodes; j++) {
        wc.uids.push_back(in->readInt32());
        wc.tags.push_back(std::string(String8(in->readString16()).string()));
      }
      mWorkChains.push_back(wc);
    }
  }
  int dataSize = 0;
  if ((res = in->readInt32(&dataSize)) != OK) {
    ALOGE("statsd could not read data size from parcel");
    return res;
  }
  if (mTagId <= 0 || mElapsedRealTimeNs <= 0 || mWallClockTimeNs <= 0 ||
      dataSize <= 0) {
    ALOGE("statsd received invalid parcel");
    return BAD_VALUE;
  }

  for (int i = 0; i < dataSize; i++) {
    int type = in->readInt32();
    switch (type) {
      case StatsLogValue::INT:
        mElements.push_back(StatsLogValue(in->readInt32()));
        break;
      case StatsLogValue::LONG:
        mElements.push_back(StatsLogValue(in->readInt64()));
        break;
      case StatsLogValue::STRING:
        mElements.push_back(
            StatsLogValue(std::string(String8(in->readString16()).string())));
        break;
      case StatsLogValue::FLOAT:
        mElements.push_back(StatsLogValue(in->readFloat()));
        break;
      case StatsLogValue::STORAGE:
        mElements.push_back(StatsLogValue());
        mElements.back().setType(StatsLogValue::STORAGE);
        in->readByteVector(&(mElements.back().storage_value));
        break;
      default:
        ALOGE("unrecognized data type: %d", type);
        return BAD_TYPE;
    }
  }
  return NO_ERROR;
};

} // Namespace os
} // Namespace android

File: core/java/android/os/StatsLogEventWrapper.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package android.os;

import android.util.Slog;

import java.util.ArrayList;
import java.util.List;

/**
 * Wrapper class for sending data from Android OS to StatsD.
 *
 * @hide
 */
public final class StatsLogEventWrapper implements Parcelable {
    static final boolean DEBUG = false;
    static final String TAG = "StatsLogEventWrapper";

    // Keep in sync with FieldValue.h enums
    private static final int EVENT_TYPE_UNKNOWN = 0;
    private static final int EVENT_TYPE_INT = 1; /* int32_t */
    private static final int EVENT_TYPE_LONG = 2; /* int64_t */
    private static final int EVENT_TYPE_FLOAT = 3;
    private static final int EVENT_TYPE_DOUBLE = 4;
    private static final int EVENT_TYPE_STRING = 5;
    private static final int EVENT_TYPE_STORAGE = 6;

    List<Integer> mTypes = new ArrayList<>();
    List<Object> mValues = new ArrayList<>();
    int mTag;
    long mElapsedTimeNs;
    long mWallClockTimeNs;
    WorkSource mWorkSource = null;

    public StatsLogEventWrapper(int tag, long elapsedTimeNs, long wallClockTimeNs) {
        this.mTag = tag;
        this.mElapsedTimeNs = elapsedTimeNs;
        this.mWallClockTimeNs = wallClockTimeNs;
    }

    /**
     * Boilerplate for Parcel.
     */
    public static final Parcelable.Creator<StatsLogEventWrapper> CREATOR = new
            Parcelable.Creator<StatsLogEventWrapper>() {
                public StatsLogEventWrapper createFromParcel(Parcel in) {
                    return new StatsLogEventWrapper(in);
                }

                public StatsLogEventWrapper[] newArray(int size) {
                    return new StatsLogEventWrapper[size];
                }
            };

    private StatsLogEventWrapper(Parcel in) {
        readFromParcel(in);
    }

    /**
     * Set work source if any.
     */
    public void setWorkSource(WorkSource ws) {
        if (ws.getWorkChains() == null || ws.getWorkChains().size() == 0) {
            Slog.w(TAG, "Empty worksource!");
            return;
        }
        mWorkSource = ws;
    }

    /**
     * Write a int value.
     */
    public void writeInt(int val) {
        mTypes.add(EVENT_TYPE_INT);
        mValues.add(val);
    }

    /**
     * Write a long value.
     */
    public void writeLong(long val) {
        mTypes.add(EVENT_TYPE_LONG);
        mValues.add(val);
    }

    /**
     * Write a string value.
     */
    public void writeString(String val) {
        mTypes.add(EVENT_TYPE_STRING);
        // use empty string for null
        mValues.add(val == null ? "" : val);
    }

    /**
     * Write a float value.
     */
    public void writeFloat(float val) {
        mTypes.add(EVENT_TYPE_FLOAT);
        mValues.add(val);
    }

    /**
     * Write a storage value.
     */
    public void writeStorage(byte[] val) {
        mTypes.add(EVENT_TYPE_STORAGE);
        mValues.add(val);
    }

    /**
     * Write a boolean value.
     */
    public void writeBoolean(boolean val) {
        mTypes.add(EVENT_TYPE_INT);
        mValues.add(val ? 1 : 0);
    }

    public void writeToParcel(Parcel out, int flags) {
        if (DEBUG) {
            Slog.d(TAG,
                    "Writing " + mTag + " " + mElapsedTimeNs + " " + mWallClockTimeNs + " and "
                            + mTypes.size() + " elements.");
        }
        out.writeInt(mTag);
        out.writeLong(mElapsedTimeNs);
        out.writeLong(mWallClockTimeNs);
        if (mWorkSource != null) {
            ArrayList<android.os.WorkSource.WorkChain> workChains = mWorkSource.getWorkChains();
            // number of chains
            out.writeInt(workChains.size());
            for (int i = 0; i < workChains.size(); i++) {
                android.os.WorkSource.WorkChain wc = workChains.get(i);
                if (wc.getSize() == 0) {
                    Slog.w(TAG, "Empty work chain.");
                    out.writeInt(0);
                    continue;
                }
                if (wc.getUids().length != wc.getTags().length
                        || wc.getUids().length != wc.getSize()) {
                    Slog.w(TAG, "Malformated work chain.");
                    out.writeInt(0);
                    continue;
                }
                // number of nodes
                out.writeInt(wc.getSize());
                for (int j = 0; j < wc.getSize(); j++) {
                    out.writeInt(wc.getUids()[j]);
                    out.writeString(wc.getTags()[j] == null ? "" : wc.getTags()[j]);
                }
            }
        } else {
            // no chains
            out.writeInt(0);
        }
        out.writeInt(mTypes.size());
        for (int i = 0; i < mTypes.size(); i++) {
            out.writeInt(mTypes.get(i));
            switch (mTypes.get(i)) {
                case EVENT_TYPE_INT:
                    out.writeInt((int) mValues.get(i));
                    break;
                case EVENT_TYPE_LONG:
                    out.writeLong((long) mValues.get(i));
                    break;
                case EVENT_TYPE_FLOAT:
                    out.writeFloat((float) mValues.get(i));
                    break;
                case EVENT_TYPE_DOUBLE:
                    out.writeDouble((double) mValues.get(i));
                    break;
                case EVENT_TYPE_STRING:
                    out.writeString((String) mValues.get(i));
                    break;
                case EVENT_TYPE_STORAGE:
                    out.writeByteArray((byte[]) mValues.get(i));
                    break;
                default:
                    break;
            }
        }
    }

    /**
     * Reads from parcel and appropriately fills member fields.
     */
    public void readFromParcel(Parcel in) {
        mTypes = new ArrayList<>();
        mValues = new ArrayList<>();
        mWorkSource = null;

        mTag = in.readInt();
        mElapsedTimeNs = in.readLong();
        mWallClockTimeNs = in.readLong();

        // Clear any data.
        if (DEBUG) {
            Slog.d(TAG, "Reading " + mTag + " " + mElapsedTimeNs + " " + mWallClockTimeNs);
        }
        // Set up worksource if present.
        int numWorkChains = in.readInt();
        if (numWorkChains > 0) {
            mWorkSource = new WorkSource();
            for (int i = 0; i < numWorkChains; i++) {
                android.os.WorkSource.WorkChain workChain = mWorkSource.createWorkChain();
                int workChainSize = in.readInt();
                for (int j = 0; j < workChainSize; j++) {
                    int uid = in.readInt();
                    String tag = in.readString();
                    workChain.addNode(uid, tag);
                }
            }
        }

        // Do the rest of the types.
        int numTypes = in.readInt();
        if (DEBUG) {
            Slog.d(TAG, "Reading " + numTypes + " elements");
        }
        for (int i = 0; i < numTypes; i++) {
            int type = in.readInt();
            mTypes.add(type);
            switch (type) {
                case EVENT_TYPE_INT:
                    mValues.add(in.readInt());
                    break;
                case EVENT_TYPE_LONG:
                    mValues.add(in.readLong());
                    break;
                case EVENT_TYPE_FLOAT:
                    mValues.add(in.readFloat());
                    break;
                case EVENT_TYPE_DOUBLE:
                    mValues.add(in.readDouble());
                    break;
                case EVENT_TYPE_STRING:
                    mValues.add(in.readString());
                    break;
                case EVENT_TYPE_STORAGE:
                    mValues.add(in.createByteArray());
                    break;
                default:
                    break;
            }
        }
    }

    /**
     * Boilerplate for Parcel.
     */
    public int describeContents() {
        return 0;
    }
}
